Source/Core/Core/IOS/Network/IP/Top.cpp

```diff
@@ -163,50 +168,170 @@ static s32 MapWiiSockOptNameToNative(u32 optname)
   {
     forward_table =
         std::unique_ptr<MIB_IPFORWARDTABLE>((PMIB_IPFORWARDTABLE) operator new(forward_table_size));
   }
 
-  std::unique_ptr<MIB_IPADDRTABLE> ip_table;
-  DWORD ip_table_size = 0;
-  if (GetIpAddrTable(nullptr, &ip_table_size, FALSE) == ERROR_INSUFFICIENT_BUFFER)
-  {
-    ip_table = std::unique_ptr<MIB_IPADDRTABLE>((PMIB_IPADDRTABLE) operator new(ip_table_size));
-  }
-
-  // find the interface IP used for the default route and use that
-  NET_IFINDEX ifIndex = NET_IFINDEX_UNSPECIFIED;
   DWORD result = GetIpForwardTable(forward_table.get(), &forward_table_size, FALSE);
   // can return ERROR_MORE_DATA on XP even after the first call
   while (result == NO_ERROR || result == ERROR_MORE_DATA)
   {
     for (DWORD i = 0; i < forward_table->dwNumEntries; ++i)
+      routing_table.push_back(
+          {forward_table->table[i].dwForwardIfIndex, forward_table->table[i].dwForwardDest,
+           forward_table->table[i].dwForwardMask, forward_table->table[i].dwForwardNextHop});
+
+    if (result == NO_ERROR)
+      break;
+
+    result = GetIpForwardTable(forward_table.get(), &forward_table_size, FALSE);
+  }
+
+#elif defined(__linux__)
+  const int buff_size = 8192;
+  unsigned int msgSeq = 0;
+  int sock = socket(PF_NETLINK, SOCK_DGRAM, NETLINK_ROUTE);
+  if (sock < 0)
+    return {};
+
+  Common::ScopeGuard socket_guard{[sock] { close(sock); }};
+  char msgbuf[buff_size];
+  memset(msgbuf, 0, buff_size);
+  auto nlMsg = (struct nlmsghdr*)msgbuf;
+  auto rtMsg = (struct rtmsg*)NLMSG_DATA(nlMsg);
+  unsigned int pid = getpid();
+
+  // prepare command/netlink packet
+  nlMsg->nlmsg_len = NLMSG_LENGTH(sizeof(struct rtmsg));  // Length of message.
+  nlMsg->nlmsg_type = RTM_GETROUTE;  // Get the routes from kernel routing table .
+
+  nlMsg->nlmsg_flags = NLM_F_DUMP | NLM_F_REQUEST;  // The message is a request for dump.
+  nlMsg->nlmsg_seq = msgSeq++;                      // Sequence of the message packet.
+  nlMsg->nlmsg_pid = pid;                           // PID of process sending the request.
+
+  // ship it
+  if (send(sock, nlMsg, nlMsg->nlmsg_len, 0) < 0)
+  {
+    return {};
+  }
+
+  // read response
+  int readlen, msglen = 0;
+  memset(msgbuf, 0, buff_size);
+
+  do
+  {
+    auto bufPtr = msgbuf + msglen;
+    readlen = recv(sock, bufPtr, buff_size - msglen, 0);
+    if (readlen < 0)
+    {
+      return {};
+    }
+
+    nlMsg = (struct nlmsghdr*)bufPtr;
+    if (NLMSG_OK(nlMsg, readlen) == 0 || nlMsg->nlmsg_type == NLMSG_ERROR)
+    {
+      return {};
+    }
+
+    // if its the last msg, we are done
+    if (nlMsg->nlmsg_type == NLMSG_DONE)
     {
-      if (forward_table->table[i].dwForwardDest == 0)
+      break;
+    }
+
+    msglen += readlen;
+    if ((nlMsg->nlmsg_flags & NLM_F_MULTI) == 0)
+    {
+      break;
+    }
+
+  } while (nlMsg->nlmsg_seq != msgSeq || nlMsg->nlmsg_pid != pid);
+
+  // parse the response
+  nlMsg = (struct nlmsghdr*)msgbuf;
```

Comment:
C++ cast should be preferred.

Reply:
unfamiliar with them, and what code what cast generates. this would be a `const_cast` ?

Reply:
I think `reinterpret_cast<nlmsghdr*>(msg_buf)` is more suited instead of these C style cast.